pkgs/development/python-modules/basedmypy/default.nix

```diff
@@ -0,0 +1,133 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  nix-update-script,
+  pythonAtLeast,
+  pythonOlder,
+  stdenv,
+
+  # build-system
+  setuptools,
+  types-psutil,
+  types-setuptools,
+  wheel,
+
+  # propagates
+  basedtyping,
+  mypy-extensions,
+  tomli,
+  typing-extensions,
+
+  # optionals
+  lxml,
+  psutil,
+
+  # tests
+  attrs,
+  filelock,
+  pytest-xdist,
+  pytestCheckHook,
+}:
+
+buildPythonPackage rec {
+  pname = "basedmypy";
+  version = "2.8.1";
+  pyproject = true;
+
+  disabled = pythonOlder "3.8";
+
+  src = fetchFromGitHub {
+    owner = "KotlinIsland";
+    repo = "basedmypy";
+    tag = "v${version}";
+    hash = "sha256-scPIcUoay8cChiKNhaXcKjN5S5G7teGCakkaFMmAJlo=";
+  };
+
+  patches = [ ./unpin-types-setuptools.patch ];
+
+  build-system = [
+    basedtyping
+    mypy-extensions
+    setuptools
+    types-psutil
+    types-setuptools
+    typing-extensions
+    wheel
+  ] ++ lib.optionals (pythonOlder "3.11") [ tomli ];
+
+  dependencies = [
+    basedtyping
+    mypy-extensions
+    typing-extensions
+  ] ++ lib.optionals (pythonOlder "3.11") [ tomli ];
+
+  optional-dependencies = {
+    dmypy = [ psutil ];
+    reports = [ lxml ];
+  };
+
+  # Compile mypy with mypyc, which makes mypy about 4 times faster. The compiled
+  # version is also the default in the wheels on Pypi that include binaries.
+  # is64bit: unfortunately the build would exhaust all possible memory on i686-linux.
+  env.MYPY_USE_MYPYC = stdenv.buildPlatform.is64bit;
+
+  # when testing reduce optimisation level to reduce build time by 20%
+  env.MYPYC_OPT_LEVEL = 1;
+
+  pythonImportsCheck =
+    [
+      "mypy"
+      "mypy.api"
+      "mypy.fastparse"
+      "mypy.types"
+      "mypyc"
+      "mypyc.analysis"
+    ]
+    ++ lib.optionals (!stdenv.hostPlatform.isi686) [
+      # ImportError: cannot import name 'map_instance_to_supertype' from partially initialized module 'mypy.maptype' (most likely due to a circular import)
+      "mypy.report"
+    ];
+
+  nativeCheckInputs = [
+    attrs
+    filelock
+    pytest-xdist
+    pytestCheckHook
+    setuptools
+    tomli
+  ] ++ lib.flatten (lib.attrValues optional-dependencies);
+
+  disabledTests =
+    [ ]
+    ++ lib.optionals (pythonAtLeast "3.12") [
+      # requires distutils
+      "test_c_unit_test"
```

Comment:
Why not add distutils to fix this?

Reply:
If I add distutils to `nativeCheckInputs`, build crashes with `Skipping analyzing "distutils": module is installed, but missing library stubs or py.typed marker.` (`distutils` stubs are in `types-setuptools` which should already be installed) during the `buildPhase`